### ISSUE TITLE
Navigation: Back button missing on deeplink/reload (3.5.0)

### ISSUE THREAD

**Issue by FdezRomero** [ionic-team/ionic#12212]
**Ionic version:**  (check one with "x")
[ ] **1.x** (For Ionic 1.x issues, please use https://github.com/ionic-team/ionic-v1)
[ ] **2.x**
[x] **3.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
In an Ionic 3.5.0 app with lazy loading, the back button is missing after following a subpage deeplink or reloading. Reproducible in the `simple-tabs` navigation test.

**Expected behavior:**
The back button to be shown.

**Steps to reproduce:**
1. Clone the `ionic-angular` repo and `npm install`.
2. Run `gulp e2e.watch --folder nav/simple-tabs`.
3. In "Page One", click the "Go to next page" button. "Page Two" is opened and the back button is shown.
4. Reload "Page Two". The back button is missing.

**Related code:**
N/A

**Other information:**
N/A

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
global packages:

    @ionic/cli-utils : 1.4.0
    Cordova CLI      : 7.0.1
    Ionic CLI        : 3.4.0

local packages:

    @ionic/app-scripts              : 1.3.12
    @ionic/cli-plugin-cordova       : 1.4.0
    @ionic/cli-plugin-ionic-angular : 1.3.1
    Cordova Platforms               : android 6.2.3 ios 4.4.0
    Ionic Framework                 : ionic-angular 3.5.0

System:

    Node       : v6.10.3
    OS         : macOS Sierra
    Xcode      : Xcode 8.3.3 Build version 8E3004b
    ios-deploy : 1.9.1
    ios-sim    : 6.0.0
    npm        : 5.0.4
```

**Comment by danbucholtz** [comment]
@FdezRomero,

This is by design. It was _really_ easy to get into a weird state of history when pushing pages on when using the browser buttons. Sometimes when you would hit `back`, you'd end up pushing a view on. We decided that calling `setRoot` and eliminating the confusion was better when reloading or visiting a page via the url that does not exist in the history. I gave some examples of the weirdness to the team, and collectively we decided this made more sense.

Having said that, we're flexible and recognize that this is a change so we will listen to community feedback on the topic and will adjust as needed. 

Thanks,
Dan

**Comment by FdezRomero** [comment]
Hi @danbucholtz,

Thanks for your reply. If I understood it right, if you open a deeplink with multiple page segments, only the last segments are used and the associated view is created as root. No history is created, and that's why the back button doesn't appear. In other words, the `defaultHistory` in `@IonicPage` is no longer used.

If this is the case, I have to disagree with this approach. Deeplinks have all the steps taken by the user so history can be recreated, and I (maybe mistakenly) thought this was the whole point of breaking with the old URLs in favor of full, improved and robust ones (in addition to supporting split views, nested nav, etc). Like they were supposed to work. Otherwise, they're... lastlinks 😉

Here goes an example: Imagine you are looking for something on Google in your Android phone and you find an app that looks interesting. You tap on the deeplink and the app details open in the Play Store. But then you don't like the app and want to look for another one. You would normally tap the back button, find the search bar there and continue looking. It's a common pattern.

But with this approach, you wouldn't have this back button in your Ionic App. You would have to kill the app from the recents screen, find the app in the homescreen and reopen it. Every time you open a deeplink. Yes, you can navigate forward inside the app, but never back.

Going back to Ionic dev mode, if you choose to go around this limitation, you could put a `hideBackButton` in the `ion-navbar`, place a button with a back arrow and manually:

1. Insert your app's main page in the history before the current view and do a nav pop (what the `defaultHistory` did) or:
2. Do a `setRoot` with your app's main page with a backwards transition.

But what's the point of pushing all the segments in the URL if they're never used? Notice that I have to go back to the main page because I don't know which segments/views I should go back to, only the Deeplinker knows. What if I don't want to force manually going back to the main page, but to its category page before the details page? I can't.

Sorry for the long comment, I'm just trying to show that I can't think of any use of deeplinks where you don't want the app to behave consistently and don't have the ability to go back. And if you do, you can just use `hideBackButton` as we're used to.

I'm willing to chat about this in more depth with you and the rest of the team if you want 👍

Thanks,
Rodrigo

**Comment by danbucholtz** [comment]
I'm not sure I'm following. Is `defaultHistory` not working for you?

Here is what happens with a navigation event directly from the URL:

1. If the page is already in the nav stack, pop to that page. This makes sense and preserves the remaining nav stack.

2. If the page is not in the nav stack, we used to `push` it on the stack. But this is a weird behavior when hitting the browser's back button. You're really `popping` in the browser, but `pushing` in the Ionic app, so the app gets into a weird state. With 3.5.0, we played around with a few scenarios as a team and decided that `setRoot` was the better behavior when a page was NOT in the nav stack.

So let's say you navigate to `page 1`, then `page 2` and `page 3` using the app's navigation. Then hit the refresh button. At this point, you refreshed the page so you don't have any history in the stack (unless you specified defaultHistory). If you hit the browser's page button from `page 3` after a refresh, it seems weird to push a page, so calling `setRoot` on `page 2` seems like a decent option.

Happy to discuss further. I'm not 100% sure I'm following what you mean.

Thanks,
Dan

**Comment by FdezRomero** [comment]
Hi Dan!

By your explanation about using `setRoot` and not pushing the history, I (mistakenly) understood that now Ionic was going to set the current page as root and that was it, no way to go back "by design". Sorry about that.

I now understand that it's supposed to keep using the `defaultHistory` to go back, but these views will be created but using `setRoot` instead of pushing the page onto the nav stack and going backwards. Makes a lot of sense. 

Then the problem here is either that `defaultHistory` isn't working or simply that Ionic's UI back button isn't shown automatically in `<ion-navbar>` like it should.

How to reproduce:
1. In `simple-tabs`, open `pages/second-page/second-page.ts` and add `defaultHistory: ['FirstPage']` to `@IonicPage`.
2. Run `gulp e2e.watch --folder nav/simple-tabs`, go to the second page and reload.
3. No back button is shown in the app's navbar, even though `FirstPage` is in the `defaultHistory`.

I think we're now on the same page, I hope this clarifies the problem about users not being able to go back.

Thanks,
Rodrigo

**Comment by danbucholtz** [comment]
Yep, looks like there is a bug. I'll take a look.

Thanks,
Dan

**Comment by FdezRomero** [comment]
Good news! 🎉

We found the root cause of the problem and have a fix working. We'll check if it passes the tests tomorrow morning and submit a PR 💪😁

Dan, can you reopen this issue?

Thanks,
Rodrigo

**Comment by danbucholtz** [comment]
@FdezRomero,

This is fixed! 🎉  🌮 

Can you verify for me? Shoot me a msg on slack if it's not. `danb` is my handle.

```
npm install ionic-angular@3.7.1-201710060319
```

Thanks,
Dan

**Comment by c0bra** [comment]
@danbucholtz this doesn't appear to be fixed in ionic 3.9.2.

I'm using defaultHistory but do not get any back button when going to a page via its segment.

**Comment by fcarreno** [comment]
Same here. I ended up writing something like this:

```
if(!this.navCtrl.canGoBack()){        
this.navCtrl.insert(0, 'TabsPage').then( () => {
           this.navBar._elementRef.nativeElement.children[1].classList.add('show-back-button');
       });
}
```
